music21/opensheetmusicdisplay/osmd.py

```diff
@@ -0,0 +1,189 @@
+# -*- coding: utf-8 -*-
+#-------------------------------------------------------------------------------
+# Name:         opensheetmusicdisplay/osmd.py
+# Purpose:      music21 classes for displaying music21 scores in IPython notebooks
+#
+# Authors:      Sven Hollowell
+#
+# Copyright:    Copyright © 2012-14 Michael Scott Cuthbert and the music21 Project
+# License:      LGPL or BSD, see license.txt
+#-------------------------------------------------------------------------------
+'''
+Registers the .show('osmd') converter for use in IPython browser
+See OSMD project page here: https://github.com/opensheetmusicdisplay/opensheetmusicdisplay
+'''
+import unittest
+from music21 import exceptions21
+
+
+class OpenSheetMusicDisplayException(exceptions21.Music21Exception):
+    pass
+
+hasInstalledIPython = False
+try:
+    # raise ImportError('for testing')
+    from IPython.core.display import display, HTML, Javascript
+    hasInstalledIPython = True
+except ImportError as e:
+    # raise error only if methods are called by a .show()
+    def display(*args, **kwargs):
+        raise OpenSheetMusicDisplayException('OpenSheetMusicDisplay requires IPython to be installed')
+    HTML = Javascript = display
+
+
+from music21.converter.subConverters import SubConverter
+from music21.instrument import Piano
+from urllib.request import pathname2url
+import time, random, json, os
+
+
+class ConverterOpenSheetMusicDisplay(SubConverter):
+    '''
+    Converter for displaying sheet music inline
+    in an IPython notebook.
+    Uses: https://opensheetmusicdisplay.org/ for rendering
+    '''
+    registerFormats = ('osmd',)
+    registerShowFormats = ('osmd',) 
+
+    def show(self, obj, fmt, 
+            fixPartName=True, offline=True, div_id=None,
+            **keywords):
+        score = obj
+        if fixPartName:
+            score = self.addDefaultPartName(score)
+
+        if div_id is None:
+            # create unique reference to output div in case we wish to update it
+            div_id = self.getUniqueDivId()
+            # div contents should be replaced by rendering
+            display(HTML('<div id="'+div_id+'">loading OpenSheetMusicDisplay</div>'))
+
+
+        xml = open(score.write('musicxml')).read()
+        script = self.musicXMLToScript(xml, div_id, offline=offline)
+
+        display(Javascript(script))
+        return div_id
+
+    @staticmethod
+    def getUniqueDivId():
+        return "OSMD-div-"+ \
+                str(random.randint(0,1000000))+ \
+                "-"+str(time.time()).replace('.','-') # '.' is the class selector
+
+    @staticmethod
+    def musicXMLToScript(xml, div_id, offline=True):
+        
+        # print('xml length:', len(xml))
+        script_dir = os.path.join(os.path.dirname(__file__), 'opensheetmusicdisplay.0.3.1.min.js')
+
+        # script that will replace div contents with OSMD display
+        script = """
```

Comment:
if the whole script tag is this extensive, maybe it should be a separate file and something to load in, so that we can run javascript checkers/linters on it?